### PR TITLE
feat(eslint-plugin): Check 'rest' parameters in no-misused-promises

### LINKED ISSUE #4015: [no-misused-promises] checksVoidReturn=true doesn't catch the error when variadic arguments are used (like `...handlers: Array<() => void>`)
This issue affects e.g. express used with new route handlers which are required to be non-async in their TS typings. I think in case of variadic arguments, the plugin verifies just the 1st argument and, if it's okay, doesn't even check the rest of the arguments.

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
     "@typescript-eslint/no-misused-promises": "error",
  }
}
```

```TS
function get2(_h1: () => void, _h2: () => void) {}

function get(..._handlers: Array<() => void>) {}

get2(
  () => {},
  async () => {} // <-- fails, which is what we want
);

get(
  () => {},
  async () => {}  // <-- this must fail too, but it does not
);
```

**Expected Result**

Both get() and get2() calls must raise an eslint error.

**Actual Result**

Only get2() is treated as incorrect, and get() is fine - although it should fail too:

image

**Additional Info**

**Versions**

| package                                   | version |
| ----------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.0.0` |
| `@typescript-eslint/parser`               | `5.0.0` |
| `TypeScript`                              | `4.4.4` |
| `ESLint`                                  | `8.0.1` |
| `node`                                    | `16.11.1` |

### PR DESCRIPTION
**Pull request by Aaron1011**
Fixes #4015

This extends 'no-misued-promises' with support for checking variadic arguments passed to a 'rest' parameter. If a function is declared with an argument like '...handlers: Array<() => void>', we now check if the type argument to `Array` is a void-returning function, and if so, check if any of the variadic arguments return a Promise.

## PR Checklist

- [X] Addresses an existing open issue: fixes  #4015
- [X] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [X] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

### CODE REVIEW

File: packages/eslint-plugin/src/rules/no-misused-promises.ts [hunk 1]
```diff
@@ -486,10 +486,13 @@ function isFunctionParam(
   return false;
 }
 
-// Get the positions of parameters which are void functions (and not also
+// Get the positions of arguments which are void functions (and not also
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-misused-promises.ts:489]: Aha, yes, good correction. Thanks 😄

File: packages/eslint-plugin/src/rules/no-misused-promises.ts [hunk 2]
```diff
@@ -507,17 +510,41 @@ function voidFunctionParams(
...
-        const type = checker.getTypeOfSymbolAtLocation(
+        const decl = parameter.getDeclarations()?.[0];
+        let type = checker.getTypeOfSymbolAtLocation(
           parameter,
           node.expression,
         );
+
+        const indices = [index];
+        // If this is a array 'rest' parameter, add all of the remaining parameter indices.
+        // Note - we currently do not support 'spread' arguments
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-misused-promises.ts:521]: Could you file an issue to support spread arguments, and link it in the code comment please? It's easy to lose track of these things, but +1 to supporting them eventually.

File: packages/eslint-plugin/src/rules/no-misused-promises.ts [hunk 3]
```diff
@@ -507,17 +510,41 @@ function voidFunctionParams(
       : subType.getConstructSignatures();
     for (const signature of signatures) {
       for (const [index, parameter] of signature.parameters.entries()) {
-        const type = checker.getTypeOfSymbolAtLocation(
+        const decl = parameter.getDeclarations()?.[0];
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-misused-promises.ts:513]: This is a bit more direct. I also got a little confused reading this about why we'd consider there to be more than one declaration.

```suggestion
        const decl = parameter.valueDeclaration;
```

File: packages/eslint-plugin/tests/rules/no-misused-promises.test.ts [hunk 4]
```diff
@@ -970,5 +970,29 @@ console.log({ ...(condition ? Promise.resolve({ key: 42 }) : {}) });
...
+);
+
+restUnion('Testing', false, () => Promise.resolve(true));
+      `,
+      errors: [
+        { line: 9, messageId: 'voidReturnArgument' },
+        { line: 10, messageId: 'voidReturnArgument' },
+        { line: 12, messageId: 'voidReturnArgument' },
+        { line: 15, messageId: 'voidReturnArgument' },
+      ],
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/no-misused-promises.test.ts:995]: We generally try to keep to one type of test case per actual test case. This looks like it tests both a `restPromises` function and a `restUnion` function. Could you please split these into two cases?

File: packages/eslint-plugin/tests/rules/no-misused-promises.test.ts [hunk 5]
```diff
@@ -970,5 +970,29 @@ console.log({ ...(condition ? Promise.resolve({ key: 42 }) : {}) });
...
+    },
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/no-misused-promises.test.ts:996]: These are very good tests to start, but I think we'll want a few more for more edge cases around tuples: `[]`, `[() => void]`, `[undefined, () => void]`, etc.

There should also be some `valid` cases added too, to make sure the rule isn't giving false reports. `Array<() => Promise>` comes to mind there.

File: packages/eslint-plugin/src/rules/no-misused-promises.ts [hunk 6]
```diff
@@ -507,17 +510,41 @@ function voidFunctionParams(
...
         );
+
+        const indices = [index];
+        // If this is a array 'rest' parameter, add all of the remaining parameter indices.
+        // Note - we currently do not support 'spread' arguments
+        if (
+          decl &&
+          ts.isParameter(decl) &&
+          decl.dotDotDotToken &&
+          checker.isArrayType(type) &&
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-misused-promises.ts:526]: If I remove this line, the tests still pass (despite the type checking complaint on `checker.getTypeArguments(type)` below). Could you add test coverage for rest param type `any`, `any[]`, and `unknown[]`?

File: packages/eslint-plugin/src/rules/no-misused-promises.ts [hunk 7]
```diff
@@ -507,17 +510,41 @@ function voidFunctionParams(
...
+          node.arguments
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-misused-promises.ts:527]: Similar note on removing this check not failing the tests now. In what situation would `node.arguments` not exist? Can we test for it?

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Looks like a great start, thanks for sending this in! Mostly requesting changes around adding unit test coverage for more edge cases, in particular tuples. Cheers!

File: packages/eslint-plugin/src/rules/no-misused-promises.ts (hunk 2 above)
Comment (Aaron1011) [reply packages/eslint-plugin/src/rules/no-misused-promises.ts:521]: Opened https://github.com/typescript-eslint/typescript-eslint/issues/5744

File: packages/eslint-plugin/src/rules/no-misused-promises.ts (hunk 3 above)
Comment (Aaron1011) [reply packages/eslint-plugin/src/rules/no-misused-promises.ts]: I've adjusted this to use `valueDeclaration`

File: packages/eslint-plugin/tests/rules/no-misused-promises.test.ts (hunk 4 above)
Comment (Aaron1011) [reply packages/eslint-plugin/tests/rules/no-misused-promises.test.ts:1064]: I've split out the union code into a separate test case

File: packages/eslint-plugin/tests/rules/no-misused-promises.test.ts (hunk 5 above)
Comment (Aaron1011) [reply packages/eslint-plugin/tests/rules/no-misused-promises.test.ts:1065]: I've added some valid tests (as well as a valid test for `new MyClass(() => Promise.resolve(1))`)

File: packages/eslint-plugin/src/rules/no-misused-promises.ts (hunk 6 above)
Comment (Aaron1011) [reply packages/eslint-plugin/src/rules/no-misused-promises.ts]: I've added some additional rest parameter tests. It turns out that we weren't handling tuples at all - I've refactored the code to add support for tuple rest arguments, and added additional tests.

File: packages/eslint-plugin/src/rules/no-misused-promises.ts (hunk 7 above)
Comment (Aaron1011) [reply packages/eslint-plugin/src/rules/no-misused-promises.ts]: This occurs when invoking a constructor without providing parameters - e.g. `let a = new MyClass;`. I've added a single check at the top of the function, which just bails out since there's nothing to check.

**Comment by Aaron1011** [comment]
@JoshuaKGoldberg Thanks for the fast review! I've pushed a new commit that should address all of your comments.

File: packages/eslint-plugin/src/rules/no-misused-promises.ts [hunk 8]
```diff
@@ -507,17 +534,59 @@ function voidFunctionParams(
...
+                node,
+                typeArgs[i - index],
+                i,
+                thenableReturnIndices,
+                voidReturnIndices,
+              );
+            }
+          } else {
+            // We must have an 'any' parameter (e.g. `...args: any`), as anything
+            // else would be invalid TypeScript. There's nothing for us to check.
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-misused-promises.ts:579]: Nit: I think this is implied by the code around it. If it's neither an array type nor a tuple type, there's nothing for us to do.

```suggestion
```

File: packages/eslint-plugin/src/rules/no-misused-promises.ts [hunk 9]
```diff
@@ -486,13 +486,40 @@ function isFunctionParam(
   return false;
 }
 
-// Get the positions of parameters which are void functions (and not also
+function checkThenableOrVoid(
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-misused-promises.ts:489]: Nit: it's specifically checking an argument. This naming confused me for a little bit.

```suggestion
function checkThenableOrVoidArgument(
```

An alternative would be to move this function declaration to _inside_ the function `voidFunctionArguments` function. That way it wouldn't need to take in the two `Set`s as parameters. I don't have a strong preference - up to you!

File: packages/eslint-plugin/src/rules/no-misused-promises.ts [hunk 10]
```diff
@@ -486,13 +486,40 @@ function isFunctionParam(
...
+// if trailing arguments are candidates.
+function voidFunctionArguments(
   checker: ts.TypeChecker,
   node: ts.CallExpression | ts.NewExpression,
 ): Set<number> {
+  // 'new' can be used without any arguments, as in 'let b = new Object;'
+  // In this case, there are no argument positions to check, so return early.
+  if (!node.arguments) {
+    return new Set<number>();
+  }
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-misused-promises.ts:522]: Great code addition, but missing test coverage - could you add one in please?

File: packages/eslint-plugin/tests/rules/no-misused-promises.test.ts [hunk 11]
```diff
@@ -970,5 +1005,78 @@ console.log({ ...(condition ? Promise.resolve({ key: 42 }) : {}) });
...
+
+restTupleOne('My string', () => Promise.resolve(1));
+restTupleTwo(true, undefined, () => Promise.resolve(true), undefined);
+restTupleFour(
+  1,
+  () => Promise.resolve(true),
+  false,
+  () => {},
+  () => Promise.resolve(1),
+);
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/no-misused-promises.test.ts:1057]: Nit: would you be open to splitting this test up too? Any test where there are multiple functions being declared feels to me like it should be split.

One motivation for splitting them is that we only get the first failure (of potentially many) when one fails.

I won't block on test splitting up much more 😄

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
This is looking good! Just a couple non-nits blocking. I'm feeling good about this one!

File: packages/eslint-plugin/src/rules/no-misused-promises.ts (hunk 10 above)
Comment (Aaron1011) [reply packages/eslint-plugin/src/rules/no-misused-promises.ts:522]: I meant for this to be tested by `new TakeCallbacks;` and `new TakesVoidCb;` but it looks like running `yarn lint --fix` changed it. I've added `// eslint-disable-next-line @typescript-eslint/internal/plugin-test-formatting` to allow it to pass the internal lint.

File: packages/eslint-plugin/src/rules/no-misused-promises.ts (hunk 8 above)
Comment (Aaron1011) [reply packages/eslint-plugin/src/rules/no-misused-promises.ts]: I've removed this branch

File: packages/eslint-plugin/src/rules/no-misused-promises.ts (hunk 9 above)
Comment (Aaron1011) [reply packages/eslint-plugin/src/rules/no-misused-promises.ts]: I've renamed the function

File: packages/eslint-plugin/tests/rules/no-misused-promises.test.ts (hunk 11 above)
Comment (Aaron1011) [reply packages/eslint-plugin/tests/rules/no-misused-promises.test.ts:1068]: I've split this into one test per function.

**Review by JoshuaKGoldberg — APPROVED** [review]
Fantastic, thanks! 👏

**Comment by jlowcs** [comment]
Following this change, my project's linting broke. I'm seeing the following:

```
TypeError: Cannot read properties of undefined (reading 'flags')
Occurred while linting /tmp/workspace/x/y/z/index.ts:19
Rule: "@typescript-eslint/no-misused-promises"
    at Object.isUnionType (/tmp/workspace/node_modules/tsutils/typeguard/2.8/type.ts:68:18)
    at Object.unionTypeParts (/tmp/workspace/node_modules/tsutils/util/type.ts:145:12)
    at isThenableReturningFunctionType (/tmp/workspace/node_modules/@typescript-eslint/eslint-plugin/src/rules/no-misused-promises.ts:625:33)
    at checkThenableOrVoidArgument (/tmp/workspace/node_modules/@typescript-eslint/eslint-plugin/src/rules/no-misused-promises.ts:497:7)
    at voidFunctionArguments (/tmp/workspace/node_modules/@typescript-eslint/eslint-plugin/src/rules/no-misused-promises.ts:568:15)
    at checkArguments (/tmp/workspace/node_modules/@typescript-eslint/eslint-plugin/src/rules/no-misused-promises.ts:216:24)
    at ruleErrorHandler (/tmp/workspace/node_modules/eslint/lib/linter/linter.js:1114:28)
    at /tmp/workspace/node_modules/eslint/lib/linter/safe-emitter.js:45:58
    at Array.forEach (<anonymous>)
    at Object.emit (/tmp/workspace/node_modules/eslint/lib/linter/safe-emitter.js:45:38)
```

**Comment by jlowcs** [comment]
submitted an issue: https://github.com/typescript-eslint/typescript-eslint/issues/5807
